x/oracle/keeper/historic_price.go

```diff
@@ -187,6 +237,33 @@ func (k Keeper) IterateHistoricPrices(
 	}
 }
 
+// IterateHistoricMediansSinceBlock iterates over medians of a given
+// denom in the store in reverse starting from the latest median
+// until startBlock.
+// Iterator stops when exhausting the source, or when the handler returns `true`.
+func (k Keeper) IterateHistoricMediansSinceBlock(
+	ctx sdk.Context,
+	denom string,
+	startBlock uint64,
+	handler func(sdk.Dec) bool,
+) {
+	store := ctx.KVStore(k.storeKey)
+
+	// make sure we have one zero byte to correctly separate denoms
+	prefix := util.ConcatBytes(1, types.KeyPrefixMedian, []byte(denom))
+	iter := sdk.KVStoreReversePrefixIterator(store, prefix)
+	defer iter.Close()
+
+	for ; iter.Valid(); iter.Next() {
+		decProto := sdk.DecProto{}
+		k.cdc.MustUnmarshal(iter.Value(), &decProto)
+		block := types.ParseBlockFromKey(iter.Key())
+		if handler(decProto.Dec) || block <= startBlock {
+			break
+		}
```

Comment:
Let's add start block to the iterator prefix. No need to access DB records we don't need.

Reply:
No, you can iterate from a given key,

Reply:
Changed it to use KVStoreReversePrefixIteratorPaginated